File: BackTracking/Generate_Parentheses.py
# Problem Link :- https://leetcode.com/problems/generate-parentheses/
"""Given n pairs of parentheses, write a function to generate all combinations 
of well-formed parentheses."""

class Solution(object):
    def generateParenthesis(self, n):
        ans = []
        self.helper("",0,0,n,ans)
        return ans
        
        
    def helper(self,s,l,r,n,ans):
        if len(s) == 2 * n:
            ans.append(s)
            return 
        if l < n:
            self.helper(s+'(',l+1,r,n,ans)
        if r < l:
            self.helper(s+')',l,r+1,n,ans)
        

File: BackTracking/GenerateParentheses.cpp
//Problem Link:- https://leetcode.com/problems/generate-parentheses/


#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    void helper(vector<string> &ans, int left, int right, int n, string curr){
        if (curr.length() >= 2 * n){
            ans.push_back(curr);
            return;
        }
        if (left < n){
            helper(ans,left+1,right,n,curr+'(');
        }
        if (right < left)
            helper(ans,left,right+1,n,curr+')');
    
    }
    
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        helper(ans,0,0, n, "");
        return ans;
    }
};